Add some AssemblyLoadContext documentation

sdmaclea opened the pull request:
## Summary

Add some AssemblyLoadContext documentation
Fix AppDomain typo
Add System.Runtime.Loader summary

Initial attempt to fix #1129 
Fixes #785

jeffschwMSFT commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,40 @@
+
+### Usage in the runtime
+ The runtime implements two <xref:System.Runtime.Loader.AssemblyLoadContext>:
+     * The <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> represents the runtimes default binding context which is used for the framework and the application main assembly and its normal dependencies as described by its `*.deps.json` file.
```
*.deps.json is really just an implementation detail.  I like the focus that Default is the 'runtimes binding context', but I think we should exclude details about deps.json.

sdmaclea replied:
Done

swaroop-sridhar commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,40 @@
+## Remarks
+ The <xref:System.Runtime.Loader.AssemblyLoadContext> represents an instance of a load context.  Conceptually, a load context creates a scope for loading, resolving, and potentially unloading a set of assemblies.
+ 
+ <xref:System.Runtime.Loader.AssemblyLoadContext> is an abstract class. While the class is abstract, only the <xref:System.Runtime.Loader.AssemblyLoadContext.Load(System.Reflection.AssemblyName)?displayProperty=nameWithType> needs to be implemented to create a concrete class.
```
Nit: Suggestion:   ~~While the class is abstract, only~~

swaroop-sridhar commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,40 @@
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Represents an assembly load context, which represents the runtime's concept of a load binding scope. This class is abstract and must be implemented to be instantiated.</summary>
```
This and several other docs use the terms loading and binding interchangeably. But the term "load binding" is very confusing, because we bind the assembly (not the load itself) in the process of loading. That is, loading is the process of resolution and binding -- is my understanding of the terminology.

swaroop-sridhar commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,40 @@
+     * The <xref:System.Reflection.Assembly.LoadFile(System.String)?displayProperty=nameWithType> method isolates the assemblies it loads by instantiating the most basic <xref:System.Runtime.Loader.AssemblyLoadContext>.  It has a simplistic isolation scheme which loads each assembly in its own <xref:System.Runtime.Loader.AssemblyLoadContext> with no dependency resolution.
+
+### Application usage
+ An application can create its own <xref:System.Runtime.Loader.AssemblyLoadContext> to create a custom solution for advanced binding scenarios. The customization focuses on defining dependency resolution mechanisms.
```
Isolation is also a strong motivation to use ALC, and we should mention that here.

sdmaclea replied:
@swaroop-sridhar this and other suggestions done.  Thanks for the solid feedback

swaroop-sridhar commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -69,7 +101,7 @@
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
+        <summary>Gets the instance of the default <see cref="T:System.Runtime.Loader.AssemblyLoadContext" />.  The default context contains the framework and the main application assembly and its dependencies.</summary>
```
Nit: "Instance of the default ALC" is a bit confusing, since default ALC itself is an instance.

swaroop-sridhar commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -189,10 +237,33 @@
+        <remarks>
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
```
Suggestion: It's worth mentioning here that different versions of the same assembly cannot be loaded into the same ALC, but can be loaded into different ALCs.

swaroop-sridhar commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -405,8 +501,20 @@
+        <remarks>
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
```
Mention that this event can be handled even for default ALC.
The LoadUnmanagedDll() and Load() methods are pre-defined for the default and cannot be overridden.

sdmaclea commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,40 @@
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Represents an assembly load context, which represents the runtime's concept of a load binding scope. This class is abstract and must be implemented to be instantiated.</summary>
```
Done

sdmaclea commented:
I have addressed all feedback.  The openpublishbuild completed without any new warnings.  I am happy with the published results so far.

@rpetrusha PTAL when you get a chance

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,45 @@
+ 
+ <xref:System.Runtime.Loader.AssemblyLoadContext> is an abstract class. The <xref:System.Runtime.Loader.AssemblyLoadContext.Load(System.Reflection.AssemblyName)?displayProperty=nameWithType> needs to be implemented to create a concrete class.
+ 
+ The <xref:System.Runtime.Loader.AssemblyLoadContext> exist primarily exists to provide assembly isolation. It allows multiple versions of the same assembly to be loaded within a single process. It replace the isolation mechanisms provided by multiple <xref:System.AppDomain> instances in the .NET Framework.
```
Nit: Probably want to leave out one of the "exist"

vitek-karas replied:
Nit: I would write it as "provide assembly **binding** isolation" (I personally would think that "assembly isolation" means isolation of everything including instances and so on).

vitek-karas replied:
"It replace**s**"

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,45 @@
+
+### Usage in the runtime
+ The runtime implements two <xref:System.Runtime.Loader.AssemblyLoadContext>:
+     * The <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> represents the runtimes default binding context which is used for the framework and the application main assembly and its normal dependencies.
```
"runtime**'s**"

vitek-karas replied:
I would say "static dependencies" (instead of normal)

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,45 @@
+     * The <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> represents the runtimes default binding context which is used for the framework and the application main assembly and its normal dependencies.
+     * The <xref:System.Reflection.Assembly.LoadFile(System.String)?displayProperty=nameWithType> method isolates the assemblies it loads by instantiating the most basic <xref:System.Runtime.Loader.AssemblyLoadContext>.  It has a simplistic isolation scheme which loads each assembly in its own <xref:System.Runtime.Loader.AssemblyLoadContext> with no dependency resolution.
+ 
+ To implement the <xref:System.Reflection.Assembly.LoadFrom(System.String)?displayProperty=nameWithType> event, the runtime adds a <xref:System.Runtime.Loader.AssemblyLoadContext.Resolving> event handler to bind to dependent files in the same directory.
```
I assume you meant "LoadFrom method" instead of "LoadFrom event"

vitek-karas replied:
The runtime registers event handler on `AssemblyLoadContext.AssemblyResolve`. The difference is that `AssemblyResolve` is a global event (fires for every load context), whether the `Resolving` is per load context.

sdmaclea replied:
Yes I misread this.  It was confusing since `AssemblyLoadContext` does not have an `AssemblyResolve` in the API.  It actually belongs in `AppDomain` in the API.

sdmaclea replied:
Removing this paragraph as it out of context.

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,45 @@
+ 
+ Additionally the <xref:System.Runtime.Loader.AssemblyLoadContext.LoadUnmanagedDll(System.String)?displayProperty=nameWithType> virtual method allows customization of the default unmanaged assembly resolution. The default implementation is often sufficient for most typical scenarios.
+
+### Testing challenges
```
I would call this "Implementation challanges" - it's not just testing, but also the implementation itself has to be somewhat careful and has to be prepared to be called more or less at any time.

sdmaclea replied:
I am choosing "Technical Challenges" and combining the section with "Limitations"

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,45 @@
+     * When code explicitly loads assemblies
+
+### Limitations
+ The runtime's assemblies may only be loaded in the default assembly.  It is not possible to load multiple versions of the runtime in a single process.
```
What is a "runtime's assembly"? In reality we only force the 'System.Private.CoreLib' to be a singleton. All other assemblies can in theory be loaded multiple times. I think we should state this clearly here. And add a warning that loading multiple copies or different versions of the framework assemblies can lead to unexpected and hard to diagnose behavior.

sdmaclea replied:
I am taking a shot at this

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -99,11 +136,15 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="assemblyPath">The path to the assembly.</param>
+        <summary>Gets an <see cref="T:System.Reflection.AssemblyName" /> for the assemblyPath.</summary>
```
I assume the `assemblyPath` needs to be somehow "quoted" or something to format it correctly as an identifier. Probably `<paramref>`.

sdmaclea replied:
I do not think it should be a param ref in the summary.  I'll leave this to @rpetrusha to determine

sdmaclea replied:
I think I should change it to "assembly path"

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -99,11 +136,15 @@
-        <returns>To be added.</returns>
+        <param name="assemblyPath">The path to the assembly.</param>
+        <summary>Gets an <see cref="T:System.Reflection.AssemblyName" /> for the assemblyPath.</summary>
+        <returns>An object that contains the fully parsed display name for the assembly at <paramref name="assemblyPath" />.</returns>
```
"fully parsed **assembly** name"

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -159,10 +202,20 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="assemblyName">The object that describes the assembly to be loaded.</param>
```
I would say something like: "An <ref>AssemblyName</ref> instance describing the assembly to be loaded."

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -159,10 +202,20 @@
+          <format type="text/markdown"><![CDATA[
+ 
+## Remarks
+ The <xref:System.Runtime.Loader.AssemblyLoadContext.Load(System.Reflection.AssemblyName)> is an abstract method. During assembly resolution, it has the first chance to resolve the assembly name to a loaded assembly.  It can either load an assembly or return `null`.  If it returns `null`, the resolution process will continue. See <xref:System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyNam ...
```
I would remove the first sentence - it's just repeating what's already state in the summary.

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -189,10 +242,37 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="assemblyName">The object that describes the assembly to be loaded.</param>
```
Like above: An AssemblyName that describes...

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -189,10 +242,37 @@
-        <remarks>To be added.</remarks>
+        <param name="assemblyName">The object that describes the assembly to be loaded.</param>
+        <summary>Resolves and loads an assembly given its <see cref="T:System.Reflection.AssemblyName" />.</summary>
+        <returns>The loaded assembly.</returns>
```
Probably should add that the method never returns null.

sdmaclea replied:
adding ", or throws"

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
"different version" is misleading. The runtime does version comparison, if the version found is equal or higher, it will accept it, otherwise it will fail.

sdmaclea replied:
Switched it to "incompatible version"

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -220,10 +300,16 @@
+        <exception cref="T:System.IO.FileLoadException">A file that was found could not be loaded.</exception>
+        <exception cref="T:System.IO.FileNotFoundException">The <paramref name="assemblyPath" /> parameter is an empty string ("") or does not exist.</exception>
+        <exception cref="T:System.BadImageFormatException">
+          <paramref name="path" /> is not a valid assembly.</exception>
```
The paramref should point to `assemblyPath`

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -283,10 +369,14 @@
-        <param name="assembly">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="assembly">A byte array that is a COFF-based image containing an emitted assembly.</param>
```
I would not use the "emitted" here. How the stream is created is of no concern to this method.

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -283,10 +369,14 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="assembly">A byte array that is a COFF-based image containing an emitted assembly.</param>
+        <summary>Loads the assembly with a common object file format (COFF)-based image containing an emitted assembly.</summary>
```
Same as above - remove "Emitted"

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -346,10 +440,18 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="unmanagedDllName">Name of the native unmanaged DLL to be loaded without its extension</param>
```
"native" and "unmanaged" are effectively synonims in this context. I would just use "unmanaged" to follow the API naming scheme.
There's no rule how the name should look like - extension or not. In fact this can be a completley random string...
Typically it is the name without extension, but not always.

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -346,10 +440,18 @@
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="unmanagedDllName">Name of the native unmanaged DLL to be loaded without its extension</param>
+        <summary>Loads an unmanaged DLL by simple name.</summary>
```
Again - "simple name" is misleading. I would simple state "Loads an unmanaged DLL specified by the <paramref unmanagedDllName/>"

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -346,10 +440,18 @@
-        <remarks>To be added.</remarks>
+        <param name="unmanagedDllName">Name of the native unmanaged DLL to be loaded without its extension</param>
+        <summary>Loads an unmanaged DLL by simple name.</summary>
+        <returns>Pointer to the loaded DLL</returns>
```
For consideration: Maybe we should not be using the term "DLL" here, as it has typically a platform specific meaning (Windows only). Maybe we should stick to the term "library". I know the API has the "DLL" in the name, but that's something we can fix anymore.

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -377,10 +479,13 @@
-        <returns>To be added.</returns>
+        <param name="unmanagedDllPath">The path to the unmanaged DLL</param>
+        <summary>Loads an unmanaged DLL from the specified path.</summary>
+        <returns>Pointer to the loaded DLL.</returns>
```
The return value should not be treated as a pointer. It's a "handle". On Windows it's the `HMODULE`, not sure what it stands for on Linux and Mac. I know that `HMODULE` on Windows really is a pointer, but I view that only as an implementation detail. Definitely the docs for Win32 APIs around this never refer to the returned value as a pointer/address.

rpetrusha replied:
"A pointer", or "A handle"

sdmaclea replied:
Linux `dlopen` docs refer to the returned `void *` as a handle.  So `A handle` seems best.

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -405,8 +510,22 @@
       <Docs>
-        <summary>To be added.</summary>
-        <remarks>To be added.</remarks>
+        <summary>Occurs when the resolution of an assembly fails.</summary>
```
I would explicitly state that this fires when an assembly resolution **on this ALC** fails - since the event is per-context.

vitek-karas commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -405,8 +510,22 @@
+ 
+ For more information about handling events, see [Handling and Raising Events](~/docs/standard/events/index.md).
+ 
+ It is noteworthy that a handler can be added for this event on any <xref:System.Runtime.Loader.AssemblyLoadContext> including the default and anonymous contexts.
```
What is the "anonymous" context?

rpetrusha replied:
```suggestion
 It is noteworthy that a handler can be added for this event on any <xref:System.Runtime.Loader.AssemblyLoadContext>, including the default and anonymous contexts.
```

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -189,10 +242,37 @@
+
+## Remarks
+ <xref:System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(System.Reflection.AssemblyName)> loads an assembly by resolving the <xref:System.Reflection.AssemblyName>.  This triggers a full resolution. The resolution fallback sequence follows this process:
+     1. The object calls <xref:System.Runtime.Loader.AssemblyLoadContext.Load(System.Reflection.AssemblyName)>.
```
"The object" should be specified -- what object? But isn't the call from this method, in which case "This method calls..." is best.

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
there's no subject in this sentence. Is it this method that attempts to load the assembly? That's true of each of the next two points as well.

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```suggestion
 > <xref:System.IO.FileLoadException> is thrown if `assemblyRef` specifies the full assembly name, and the first assembly that matches the simple name has a different version or culture. The loader does not continue probing for other assemblies that match the simple name.
```

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```suggestion
 Each <xref:System.Runtime.Loader.AssemblyLoadContext> can load only:
```

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -220,10 +300,16 @@
+          <paramref name="path" /> is not a valid assembly.</exception>
       </Docs>
     </Member>
     <Member MemberName="LoadFromNativeImagePath">
```
Can you document LoadFromNativeImagePath as well?

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -346,10 +440,18 @@
-        <remarks>To be added.</remarks>
+        <param name="unmanagedDllName">Name of the native unmanaged DLL to be loaded without its extension</param>
+        <summary>Loads an unmanaged DLL by simple name.</summary>
+        <returns>Pointer to the loaded DLL</returns>
```
```suggestion
        <returns>A pointer to the loaded DLL</returns>
```

rpetrusha replied:
But "A handle" instead?

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -346,10 +440,18 @@
+ ]]></format>
+        <exception cref="T:System.ArgumentNullException">
+          <paramref name="unmanagedDllName" /> is <see langword="null" />.</exception>
+        </remarks>
```
The \<remarks> element must precede the \<exception> element.

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -346,10 +440,18 @@
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
+ This virtual method can be overridden to customize the unmanaged DLL search algorithm.
```
It would be good to elaborate on simple name with an example or two.

rpetrusha commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -405,8 +510,22 @@
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
+ It is the responsibility of the handler for this event to return the assembly that is specified or to return null if the assembly is not recognized.
```
```suggestion
 It is the responsibility of the handler for this event to return the assembly that is specified, or to return `null` if the assembly is not recognized.
```

rpetrusha commented on xml/ns-System.Runtime.Loader.xml:
```diff
@@ -1,6 +1,6 @@
 <Namespace Name="System.Runtime.Loader">
   <Docs>
-    <summary>To be added.</summary>
+    <summary>The <see cref="N:System.Runtime.Loader" /> namespace provides interfaces and classes to allow developers the opportunity to inluence runtime loading behavior.</summary>
```
```suggestion
    <summary>The <see cref="N:System.Runtime.Loader" /> namespace provides interfaces and classes to allow developers the opportunity to influence runtime loading behavior.</summary>
```

sdmaclea commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -20,8 +20,45 @@
+
+### Usage in the runtime
+ The runtime implements two <xref:System.Runtime.Loader.AssemblyLoadContext>:
+     * The <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> represents the runtimes default binding context which is used for the framework and the application main assembly and its normal dependencies.
```
"class libraries" is not as specific as I'd like.  I chose to drop the phrase as it is redundant with static dependencies.

sdmaclea commented on xml/System.Runtime.Loader/AssemblyLoadContext.xml:
```diff
@@ -189,10 +242,37 @@
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
+ <xref:System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(System.Reflection.AssemblyName)> loads an assembly by resolving the <xref:System.Reflection.AssemblyName>.  This triggers a full resolution. The resolution fallback sequence follows this process:
```
What does `%2A` mean?

sdmaclea commented:
I have tried to address all the feedback.  Waiting for the build to complete to take a look again.

sdmaclea commented:
Build looks reasonable to me.  This is the link https://review.docs.microsoft.com/en-us/dotnet/api/system.runtime.loader?view=netcore-2.2&branch=pr-en-us-1353

rpetrusha commented:
@sdmaclea, %2A means that the method name will be displayed without parameters, and that, if the method is overloaded, the link resolves to the top of the page (with the overload list) rather than to a specific overload.

rpetrusha commented:
This looks good, @sdmaclea. I'll merge your PR now. Hopefully, it will be live tomorrow.
